compiler/rustc_const_eval/src/const_eval/eval_queries.rs

```diff
@@ -119,7 +119,7 @@ pub(super) fn op_to_const<'tcx>(
     // the usual cases of extracting e.g. a `usize`, without there being a real use case for the
     // `Undef` situation.
     let try_as_immediate = match op.layout.abi {
-        Abi::Scalar(..) => true,
+        Abi::Scalar(abi::Scalar::Primitive { .. }) => true,
```

Comment:
In particular `scalar.primitive()` always works (even for unions), which muddles the idea that "Primitive implies noundef".

Reply:
I initially added the `primitive` function to make transitioning easier and change no codegen behaviour for now. I do wonder whether unions should have a primitive layout beyond size and align requirements.

Wrt naming, I didn't want to have a `NoUndef` variant due to the double negative, but... it may be the clearest due to people knowing what it means

Reply:
I think we need the `Primitive` since we want to generate the appropriate LLVM type, and it might affect the call ABI.

> Wrt naming, I didn't want to have a NoUndef variant due to the double negative, but... it may be the clearest due to people knowing what it means

It could also be `InitializedPrimitive` or so. (IMO we should not use "undef" at all, and rather use "uninit", as that is the term we usually use in Rust.)

Reply:
Or just `Initialized`, I like it. :)